#1264 Encounter modal validation improvments

CarlosNZ opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #1264

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->

Adds validation to the new Encounter modal and won't let user submit until all required fields are valid. (as per [this comment](https://github.com/openmsupply/open-msupply/issues/1264#issuecomment-1458892171) and the following one)

I added some general internal validation to the Date and Time pickers -- the problem was that the UI error state of the input wasn't accurately reflecting the validation checks done by the caller. And if we want the calling component to only update its `value` when the `onChange` value is a valid date, then the base component will never know if the parent value is valid or not. How it works now is that the picker components manage some internal state that gets updated on every `onChange`, but only when the input is a valid date does it call the parent (passed-in) `onChange`. 

(We could have also just passed down an `error` prop from above, but this way seems simpler as most of the validation can be handled by the base picker rather than having to write `isValid` checks for every implementation)

We also take advantage of the picker's built-in `onError` function to update the "formError" state of the parent whenever the picker goes into error state.

# 🧪 How has/should this change been tested? 

Start a new encounter, check that the "Create" button only becomes active once all fields (except "endTime") are entered and valid. You should notice that if you then start manually entering a date or time in a previously valid form, the "Create" button will become inactive again until a full valid date/time is entered.

Also, in the main "Encounter" page, changing the start date / time manually should similarly show error state for invalid manual entries.

## 💌 Any notes for the reviewer?

I'm still not fully happy with the interaction between the "Date" and "Time" inputs in the "New Encounter" form. (e.g. invalidating one of them resets the other once a new valid value is added). This is because there is two input fields both updating a single "startDateTime" value. If we want to improve this, I think we'll have to go with a combined single "Date/Time" picker, so let me know what seems best.

See inline comments for additional details.

CarlosNZ commented on client/packages/common/src/ui/components/inputs/TimePickers/TimePickerInput/TimePickerInput.tsx:
```diff
@@ -1,11 +1,35 @@
 export const TimePickerInput: FC<
   Omit<TimePickerProps<Date>, 'renderInput'>
 > = props => {
+  const [internalValue, setInternalValue] = useState<
+    Date | string | number | null
+  >();
+
+  useEffect(() => {
+    if (props.value) setInternalValue(props.value);
+  }, [props.value]);
+
+  const isInvalid = (value: Date | string | number | null | undefined) => {
+    const dateValue =
+      typeof value === 'string' ? DateUtils.getDateOrNull(value) : value;
+    return !!value && !DateUtils.isValid(dateValue);
+  };
+
+  const debouncedOnChange = useDebounceCallback(
+    value => {
+      if (DateUtils.isValid(value)) props.onChange(value);
+      else setInternalValue(value);
+    },
+    [props.onChange]
+  );
+
```
Basically identical to the BaseDatePicker changes

CarlosNZ commented on client/packages/system/src/Encounter/DetailView/Toolbar.tsx:
```diff
@@ -123,9 +123,7 @@ export const Toolbar: FC<ToolbarProps> = ({ encounter, onChange }) => {
-                      const startDatetime = date
-                        ? DateUtils.formatRFC3339(date)
-                        : undefined;
+                      const startDatetime = DateUtils.formatRFC3339(date);
```
No need for `date` check as this `onChange` function won't be called by the base date picker until it gets a valid date. Also, the `formatRFC3339` can handle null/undefined values anyway.

CarlosNZ commented on client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx:
```diff
@@ -72,30 +72,28 @@ export const CreateEncounterModal: FC = () => {
-        ...draft,
-        startDatetime,
-        status: DateUtils.isFuture(date)
+    setFormError(false);
+    setDraft({
+      createdDatetime,
+      ...draft,
+      startDatetime,
+      status:
+        date && DateUtils.isFuture(date)
```
Internal validation removes need for these additional checks.

CarlosNZ commented on client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx:
```diff
@@ -154,33 +155,30 @@ export const CreateEncounterModal: FC = () => {
-                      value={DateUtils.getDateOrNull(
-                        draft?.startDatetime ?? null
-                      )}
+                      value={draft?.startDatetime}
```
Moved the conversion from string to Date object into the actual Picker components -- this prevents another one of those `useEffect` errors where it thinks a value has changed but it actually hasn't, it just got a new instance of the same thing.

CarlosNZ commented:
Also, I can confirm that the internal error discussed [over here](https://github.com/openmsupply/open-msupply/pull/1279#issuecomment-1457110891) is still present. Not much I can do about that, but might be another justification for switching to a single DateTime picker input?

clemens-msupply commented:
Was able to create this:
![image](https://user-images.githubusercontent.com/88299195/224581579-6d9a3a52-0e5f-4d16-bb14-34fb65e9a405.png)

To reproduce: set to 01/00/2022 and then change to 01/01/2022

clemens-msupply commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -1,13 +1,36 @@
+      else setInternalValue(value);
+    },
+    [props.onChange]
+  );
```
There are some issues with useDebounceCallback... (

Could you please use something like to rule this out (taken from `packages/programs/src/JsonForms/common/hooks/useDebouncedTextInput.ts`):
```
const debouncedText = useDebouncedValue(text, 500);
  useEffect(() => {
    handleChange(debouncedText);
  }, [debouncedText]);
```

Note, I only realized now that there is already a `useDebouncedValue` implementation in common...

CarlosNZ replied:
Are you actually seeing any specific problems attributable to the `useDebounceCallback` in this case? Cos if so, and we consider it problematic, we should probably make a separate issue to remove it completely and replace it with some of these other things. But that would be quite a big job -- but I was under the impression that it's working fine for most use cases.

clemens-msupply replied:
yes should consider to replace it...

For the text input the issue was when entering text quickly, so yes probably an issue here as well and likely that we run in the race condition as well here.

clemens-msupply reviewed (changes requested):
Some issues I noticed during testing (Create encounter form):

- set the time and then change the date -> the time will be reset to 12:00am
- no indication what is missing, e.g. when not start date is selected
- event if all required data is entered the button stays disabled (end date is optional isn't it?):
![image](https://user-images.githubusercontent.com/88299195/224582659-fe73578e-d7f7-4290-a0db-46170977f7e4.png)
- if the start date is invalid, setting the time input values doesn't work properly (it jumps back to previous values)
- invalid data but create button is enabled: 
![image](https://user-images.githubusercontent.com/88299195/224582839-9563a9f7-0246-4d55-9a8e-404959a02c31.png)
- Was able to create this:
![image](https://user-images.githubusercontent.com/88299195/224581579-6d9a3a52-0e5f-4d16-bb14-34fb65e9a405.png)

To reproduce: set to 01/00/2022 and then change to 01/01/2022

clemens-msupply commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -1,13 +1,36 @@
+
+  useEffect(() => {
+    if (props.value) setInternalValue(props.value);
+  }, [props.value]);
```
Think you have to remember the time part from the props and later merge them back together...

CarlosNZ commented:
> set the time and then change the date -> the time will be reset to 12:00am

I think this and probably a number of the other issues are due to the fact that the Time and Date pickers are sharing a single DateTime value and they both update it independently. Unfortunately the Time picker still takes a "DateTime" as its value, so it's actually kind of tricky to separate them out. I was going to keep them as separate date and time values and just re-combine them before saving to JSONForms, but then the time picker would have to hold a "DateTime" with a "dummy" date part, which seemed a little clunky. And there were various other difficulties associated with it.

As I mentioned in my comments, I think if we want to make this more robust, we'd be better to just change the input to a single DateTime picker (like we have in JSONForms, where you have to enter both sequentially). I'm happy with this option -- are you fine to just go with that?

CarlosNZ commented:
> * no indication what is missing, e.g. when not start date is selected

We currently have no indicators for required fields. We should possibly do something about this, but would require further discussion. Have made an issue to get the ball rolling: #1343 

> * event if all required data is entered the button stays disabled (end date is optional isn't it?):

"Clinician" is also required here, that's why button is still disabled. But I agree that it's not clear -- see previous comment regarding indication required fields.

CarlosNZ commented on client/packages/common/src/intl/utils/DateUtils.ts:
```diff
@@ -86,8 +86,8 @@ export const DateUtils = {
   startOfYear,
-  formatRFC3339: (date: Date) =>
-    isValid(date) ? formatRFC3339(date) : undefined,
+  formatRFC3339: (date?: Date | null) =>
```
No argument and an argument with a value of `undefined` *should* be considered the same thing (and as far as I can tell JS treats them the same in almost all cases). So this is just a stylistic thing and the code base generally follows the `?:` syntax over `undefined` doesn't it?

clemens-msupply replied:
it depends how you want users to use your method. When having as `date: Date | null | undefined` the user must provide an argument and forgetting it is an error. In this case not providing and argument doesn't make sense and we should treat it as an error. So not a stylisitic thing but a way to avoid errors :)

clemens-msupply commented:
> > set the time and then change the date -> the time will be reset to 12:00am
> 
> I think this and probably a number of the other issues are due to the fact that the Time and Date pickers are sharing a single DateTime value and they both update it independently. Unfortunately the Time picker still takes a "DateTime" as its value, so it's actually kind of tricky to separate them out. I was going to keep them as separate date and time values and just re-combine them before saving to JSONForms, but then the time picker would have to hold a "DateTime" with a "dummy" date part, which seemed a little clunky. And there were various other difficulties associated with it.
> 
> As I mentioned in my comments, I think if we want to make this more robust, we'd be better to just change the input to a single DateTime picker (like we have in JSONForms, where you have to enter both sequentially). I'm happy with this option -- are you fine to just go with that?
> 
> (As an added bonus, the DateTime picker doesn't have the crashing bug that the Time picker has, so that would be another reason to go with it and not use the Time picker alone)

Yes agree, think a single control to enter the DateTime would easier. (Could still have single time picker for the endtime though?)

CarlosNZ commented:
> (Could still have single time picker for the endtime though?)

Yes, seems right.

CarlosNZ commented:
Have taken care of the functionality issues, @clemens-msupply and combined the start date into a single DateTime picker.

As far as I can tell it behaves correctly now. There's one little quirk that doesn't *quite* behave exactly as you might want it, but it's pretty niche, and strictly speaking it's still "correct" -- see if you can spot it (and if you can't then great!)

clemens-msupply commented on client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx:
```diff
@@ -107,14 +102,17 @@ export const CreateEncounterModal: FC = () => {
   };
 
+  const canSubmit = () =>
+    draft !== undefined && draft.clinician && draft.startDatetime;
```
```suggestion
    draft !== undefined && draft.startDatetime;
```

Think we shouldn't make this required because it might be unknown, e.g. when scheduling an encounter. And the demo server doesn't has clinicians set

CarlosNZ replied:
Sure, no longer required.

clemens-msupply reviewed (changes requested):
Date time inputs work great now!

CarlosNZ commented:
Added "Blocked" label to this issue, as I can't really finalise it until we update from `develop` so I can resolve the discrepancies with the Date Pickers (they've been upgraded in `develop`, so I'll need to make some small changes in this PR to be compatible to the fixes I added there)

Have made a Draft PR with where it's at currently, if anyone has any insight or wants to have a look? #1411

CarlosNZ commented:
> Added "Blocked" label to this issue, as I can't really finalise it until we update from `develop` so I can resolve the discrepancies with the Date Pickers (they've been upgraded in `develop`, so I'll need to make some small changes in this PR to be compatible to the fixes I added there)
> 
> Have made a Draft PR with where it's at currently, if anyone has any insight or wants to have a look? #1411

Okay @clemens-msupply, rather than try and fix all those package changes here, I've left it for now. But I *have* updated the Date and Time pickers to be consistent with how the DatePicker now works in `develop` so when we do get round to updating from develop, these should be basically okay.

One small difference between this and how I did it last time is that the pickers don't set their parent components to `null` when an invalid value is present. This means that the parent (in this case the CreateEncounterModal) holds to its last good value when an invalid value is present in the form, which means that alone is not enough to determine if the form is submittable. So I've restored a "formError" state value for each of the start and end times, which responds to the "onError" method of the pickers.

Have also made "Clinician" not required, as suggested.

28782601a

clemens-msupply commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -1,20 +1,46 @@
+
+  useEffect(() => {
+    // This sets the internal state from parent when first loading (i.e. when
+    // the internal date is still empty)
```
Nit: Not only when first loading but potentially when value changes, right? not sure if this is important for the logic though?

CarlosNZ replied:
No, it's only on first load, hence the `internalValue === null` check. Otherwise the internalValue gets set twice on userInput -- once due to local `onChange` and again after parent value changes (after debounce).

If we need it to respond to external changes in the parent value, we can address that once we understand the specific case, but for now, it's always set from user input except on first load, where it loads the saved value.

clemens-msupply commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -1,20 +1,46 @@
+    // the internal date is still empty)
+    if (value && internalValue === null)
+      setInternalValue(DateUtils.getDateOrNull(value));
+  }, [value]);
```
`internalValue` is missing from the dependency array...

CarlosNZ replied:
Yes, that is correct. If we set `internalValue` on change of `internalValue`, we'd have an infinite loop. ;)

clemens-msupply replied:
You should always include the full dependencies array if you want to avoid subtle bugs (see the pitfall box in the useEffect docs).

Why is there a infinite loop? the `if` should catch it shouldn't it?

CarlosNZ replied:
Yes, I have read about that and seen the ESLint warnings. However, I have yet to see an example of these "subtle bugs" that has convinced me to over-stuff the dependency array. In this case, why would we want to re-run the useEffect function on change of `internalValue`? All it does is (possibly) _update_ `internalValue`? And we're comparing against the initial state of `internalValue`, so it'll always be `null` on first load.

I'm open to being convinced otherwise, I just don't really see a problem in this case. To me it just adds confusion.

clemens-msupply replied:
beside its a React convention/contract and the occasional bug caused by not specifying it, it becomes IMHO also harder to read. For example, in my previous comment I was not sure if the comment was wrong or the dependency was missing intentionally. 

When you not specify the full array your code become more "concurrent" because you have to mentally keep track which state variables are in and in what order things are called. If you specify all dependencies in the array you can just systematically handle all possible combinations in useEffect which simplifies the problem a bit...

clemens-msupply commented on client/packages/common/src/ui/components/inputs/DatePickers/DateTimePickerInput/DateTimePickerInput.tsx:
```diff
@@ -0,0 +1,94 @@
+    // the internal date is still empty)
+    if (value && internalValue === null)
+      setInternalValue(DateUtils.getDateOrNull(value));
+  }, [value]);
```
same comments as above :) `internalValue` is missing from dependency array

clemens-msupply reviewed (approved):
Works great thank you!
